File: 16.R
# Power digit sum

Args  <- commandArgs()
power <- ifelse(is.na(Args[6]), 1000, as.numeric(Args[6]))

digits <- c(1)

for (i in seq(power)) {
    doubled_digits <- vector()
    carry <- 0
    for ( digit in digits ) {
        sum <- digit * 2 + carry

        # Get last digit of sum and keep carry over
        last_digit <- as.integer(substr(sum, nchar(sum), nchar(sum)))
        doubled_digits <- c(doubled_digits, last_digit)
        if ( sum >= 10 ) {
            carry <- as.integer(substr(sum, 1, nchar(sum)-1))
        } else {
            carry <- 0
        }
    }
    if ( carry > 0 ) {
        carry_digits <- as.integer(strsplit(as.character(carry), split=NULL)[[1]])
        doubled_digits <- c(doubled_digits, carry_digits)
    }
    digits = doubled_digits
}

cat(sum(digits), fill=TRUE)

File: 16.js
#!/usr/bin/env node

// Power digit sum

var program = require('commander');

program
    .version('0.1.0')
    .description('Power digit sum')
    .option('-p, --power <int>', 'The power to raise 2 to', Number, 1000)
    .parse(process.argv);

var digits = [1];

for (var i = 1; i <= program.power; i++) {
    var doubled_digits = [];
    var carry = 0;
    for (var j = 0; j < digits.length; j++) {
        var sum = digits[j] * 2 + (carry || 0);
        var last_digit_of_sum = parseInt(sum.toString().slice(-1));
        doubled_digits.push(last_digit_of_sum);
        carry = parseInt(sum.toString().slice(0, -1));
    }
    if (carry) {
        var carry_digits = carry.toString().split("").reverse();
        for (var j = 0; j < carry_digits.length; j++) {
            doubled_digits.push(parseInt(carry_digits[j]));
        }
    }
    digits = doubled_digits;
}

var digits_sum = 0;
for (var i = 0; i < digits.length; i++) {
    digits_sum += digits[i];
}

console.log(digits_sum);
